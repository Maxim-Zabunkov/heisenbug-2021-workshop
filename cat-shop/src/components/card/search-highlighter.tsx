import { createStyles, makeStyles } from "@material-ui/core";
import { yellow } from "@material-ui/core/colors";
import React from "react";
import { useSelector } from "react-redux";
import { AppState } from "../../contracts/app-state.contracts";

const useStyle = makeStyles(() =>
    createStyles({
        highlighted: {
            backgroundColor: yellow[500]
        }
    })
);

export function SearchHighlighter(props: { text: string }): JSX.Element | null {
    const { text } = props;
    const classes = useStyle();
    const search = useSelector((x: AppState) => x.search?.pattern);
    if (!search)
        return <>{text}</>;
    const parts = splitText(text, search);
    return (
        <>
            {parts.map((p, i) => <span key={i} className={p.highlight ? classes.highlighted : undefined}>{p.text}</span>)}
        </>
    );
}

export function splitText(original: string, search: string): { text: string; highlight?: true }[] {
    const parts = original.split(new RegExp(search, 'i'));
    let lastIndex = 0;
    return parts.flatMap((text, i, a) => {
        const item = { text };
        lastIndex += text.length;
        if (i < a.length - 1) {
            const highlight = { text: original.substr(lastIndex, search.length), highlight: true }
            lastIndex += search.length;
            return [item, highlight];
        }
        return [{ text }];
    });
}