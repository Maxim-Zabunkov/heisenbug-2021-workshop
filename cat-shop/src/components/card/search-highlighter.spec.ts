import { splitText } from "./search-highlighter";

describe('SerachHighlighter', () => {
    test('should split text onto parts', () => {
        const result = splitText('Hello My Friend', 'fri');
        expect(result).toMatchSnapshot();
    });
});