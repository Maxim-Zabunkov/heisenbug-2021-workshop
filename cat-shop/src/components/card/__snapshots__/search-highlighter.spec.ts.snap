// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SerachHighlighter should split text onto parts 1`] = `
Array [
  Object {
    "text": "Hello My ",
  },
  Object {
    "highlight": true,
    "text": "Fri",
  },
  Object {
    "text": "end",
  },
]
`;
